Barbara builds an async executor (#115)
rust-lang/wg-async-foundations: src/vision/status_quo/barbara_builds_an_async_executor.md

# 😱 Status quo stories: Barbara builds an async executor

## 🚧 Warning: Draft status 🚧

This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 

## The story

Barbara wants to set priorities to the tasks spawned to the executor. However, she finds no existing async executor provides such a feature so she decided to build her own async executor.

> First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. It is not the main part of the story. Each working thread can get a task from the scheduler and poll it in loop.

Reviewer: This paragraph says she _could_ use it to build a scheduler, and below the existence of the scheduler is assumed. Maybe this paragraph should say she _does_ use it to build a scheduler. Also, it could briefly name the operations of that scheduler so the reader knows what `send` is.

Reviewer 2: ```suggestion
First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She decides to use it to build task schedulers. It is not the main part of the story. She plans for each working thread to have a loop which repeatedly gets a task from the deque and polls it.
```

Author: Ah yes, we can assume that she has made a task scheduler before building the executor.

Reviewer 3: Nit: I'm not sure the sentence "It is not the main part of the story." is actually helpful. We could probably just remove it.

> But wait, what is a "task"?

Reviewer 2: ```suggestion
But wait, what should we put into those queues to represent each "task"?
```

I think this is what you mean?

Author: Exactly. Thanks for clarifying it.

At first, Barbara thought it must contain the `Future` itself and the additional priority which was used by the scheduler. So she first wrote:

```rust
pub struct Task {
    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
    priority: u8
}
```

And the working thread loop should run something like:

```rust
pub fn poll_task(task: Task) {
    let waker = todo!();
    let mut cx = Context::from_waker(&waker);
    task.future.as_mut().poll(&mut cx);
}
```

> "How to create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. After reading the documentation, she realized the task in the scheduler should be stored in an `Arc`. And to implement `Wake`, the `Task` should also contain the sender of the scheduler. She changed the code to something like this:

Reviewer: At this point I'm thinking about building an I/O reactor, so it might help to explain (here or above) that this is for compute tasks, not async I/O ones.

Reviewer 2: It'd be helpful I think to say *why* the task should be in an arc. Presumably the idea is that the `Waker` will need to hold a reference to the task so that it can be scheduled; I guess it will need to also hold a reference to the queue or something like that?

Reviewer 2: Oh, I guess you said that last part ('the sender of the scheduler")

Reviewer 2: Oh, and I guess the `Arc` is partly because of the waker API...

    let waker = Waker::from(task.clone());
    let mut cx = Context::from_waker(&waker);
    task.future.as_mut().poll(&mut cx);
//  ^^^^^^^^^^^ cannot borrow as mutable
}
```

The code still needed some change because the `future` in the  `Arc<Task>` became immutable. 

"Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>` and I think the same future will not be polled concurrently in two threads. So let me bypass the safety checks." Barbara confidently used some `unsafe` blocks to make it compile.

```rust
pub struct Task {
    future: *mut (dyn Future<Output = ()> + Send + 'static),
    ...
}

unsafe impl Send for Task {}
unsafe impl Sync for Task {}

pub fn poll_task(task: Arc<Task>) {
    ...
    unsafe {
        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
    }
}
```

Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.

> Barbara soon got a idea to solve it. She added a state field to the `Task`. By carefully maintaining the state of the task, she could guarantee there are no duplicate tasks in the scheduler and no race can happen when polling the future:

Reviewer: Just as a suggestion, maybe this paragraph can say something about how she spends a lot of effort in writing the code below (at least, it would take me a lot of effort). Then the reader is already focused on the pain of having to write this to satisfy such a simple requirement.

```rust
const NOTIFIED: u64 = 1;
const IDLE: u64 = 2;
const POLLING: u64 = 3;
const COMPLETED: u64 = 4;

pub struct Task {
    ...
    state: AtomicU64,
}

impl Wake for Task {
    fn wake(self: Arc<Self>) {
        let mut state = self.state.load(Relaxed);
        loop {
            match state {
                IDLE => match self
                    .state
                    .compare_exchange_weak(IDLE, NOTIFIED, AcqRel, Acquire)
                {
                    Ok(_) => self.scheduler.send(self.clone()),
                    Err(s) => state = s,
                },
                POLLING => match self
                    .state
                    .compare_exchange_weak(POLLING, NOTIFIED, AcqRel, Acquire)
                {
                    Ok(_) => break,
                    Err(s) => state = s,
                },
                _ => break,
            }
        }
    }
}
pub fn poll_task(task: Arc<Task>) {
    let waker = Waker::from(task.clone());
    let mut cx = Context::from_waker(&waker);
    loop {

>         task.state.store(POLLING, Release);

Reviewer: I got tripped up wondering why you didn't need to read before polling (it's because wake() prevents the task from appearing in the scheduler twice). You do state this in the prose above, but maybe a comment here would help as well.

        if let Poll::Ready(()) = unsafe { Pin::new_unchecked(&mut *task.future).poll(&mut cx) } {
            task.state.store(COMPLETED, Release);
        }
        match task.state.compare_exchange(POLLING, IDLE, AcqRel, Acquire) {
            Ok(_) => break,
            Err(NOTIFIED) => continue,
            _ => unreachable!(),
        }
    }
}
```

Barbara finished her initial implementation of the async executor. Despite there were a lot more possible optimizations, Barbara already felt it is a bit complex. She was also confused about why she needed to care so much about polling and waking while her initial requirement was just adding additional information to the task for customizing scheduling.

## 🤔 Frequently Asked Questions

*Here are some standard FAQ to get you started. Feel free to add more!*

> * **What are the morals of the story?**

Reviewer 4: The style of the FAQ sections was changed in #118 / #119, using sub headings (###) instead of nested bullet lists.

# 😱 Status quo stories: Barbara builds an async executor

## 🚧 Warning: Draft status 🚧

This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async Rust programmers face today. 

## The story

Barbara wants to set priorities to the tasks spawned to the executor. However, she finds no existing async executor provides such a feature so she decided to build her own async executor.

> First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. It is not the main part of the story. Each working thread can get a task from the scheduler and poll it in loop.

Reviewer 2: ```suggestion
First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She could use it to build task schedulers. Each working thread can get a task from the scheduler and poll it in loop.
```

> First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She decides to use it to build task schedulers. It is not the main part of the story. She plans for each working thread to have a loop which repeatedly gets a task from the deque and polls it.

Reviewer 2: ```suggestion
First, Barbara found [crossbeam-deque](https://crates.io/crates/crossbeam-deque) provides work-stealing deques of good quality. She decides to use it to build task schedulers. She plans for each working thread to have a loop which repeatedly gets a task from the deque and polls it.
```

But wait, what is a "task"?

At first, Barbara thought it must contain the `Future` itself and the additional priority which was used by the scheduler. So she first wrote:

```rust
pub struct Task {
    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
    priority: u8
}
```

And the working thread loop should run something like:

```rust
pub fn poll_task(task: Task) {
    let waker = todo!();
    let mut cx = Context::from_waker(&waker);
    task.future.as_mut().poll(&mut cx);
}
```

> "How to create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. Seeing the `wake` method takes an `Arc<Self>`, she realized the task in the scheduler should be stored in an `Arc`. After some thought, she thinks it makes sense because both the deque in the scheduler and the waker may hold a reference to the task.

Reviewer 3: ```suggestion
"How do I create a waker?" Barbara asked herself. Quickly, she found the `Wake` trait. Seeing the `wake` method takes an `Arc<Self>`, she realized the task in the scheduler should be stored in an `Arc`. After some thought, she realizes it makes sense because both the deque in the scheduler and the waker may hold a reference to the task.
```

To implement `Wake`, the `Task` should contain the sender of the scheduler. She changed the code to something like this:

```rust
pub struct Task {
    future: Pin<Box<dyn Future<Output = ()> + Send + 'static>>,
    scheduler: SchedulerSender,
    priority: u8,
}

unsafe impl Sync for Task {}

impl Wake for Task {
    fn wake(self: Arc<Self>) {
        self.scheduler.send(self.clone());
    }
}

pub fn poll_task(task: Arc<Task>) {
    let waker = Waker::from(task.clone());
    let mut cx = Context::from_waker(&waker);
    task.future.as_mut().poll(&mut cx);
//  ^^^^^^^^^^^ cannot borrow as mutable
}
```

The code still needed some change because the `future` in the  `Arc<Task>` became immutable. 

> "Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>` and I think the same future will not be polled concurrently in two threads. So let me bypass the safety checks." Barbara confidently used some `unsafe` blocks to make it compile.

Reviewer 3: ```suggestion
"Okay. I can guarantee `Task` is created from a `Pin<Box<Future>>`, and I think the same future won't be polled concurrently in two threads. So let me bypass the safety checks." Barbara changed the future to a raw pointer and confidently used some `unsafe` blocks to make it compile.
```

```rust
pub struct Task {
    future: *mut (dyn Future<Output = ()> + Send + 'static),
    ...
}

unsafe impl Send for Task {}
unsafe impl Sync for Task {}

pub fn poll_task(task: Arc<Task>) {
    ...
    unsafe {
        Pin::new_unchecked(&mut *task.future).poll(&mut cx);
    }
}
```

> Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of it and a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.

Reviewer 3: ```suggestion
Luckily, a colleague of Barbara noticed something wrong. The `wake` method could be called multiple times so multiple copies of the task could exist in the scheduler. The scheduler might not work correctly because of this. What's worse, a more severe problem was that multiple threads might get copies of the same task from the scheduler and cause a race in polling the future.
```
